T/897: Make Position/Range immutable.

jodator opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Other: Make `Position` and `Range` immutable in model and view. Closes #897.

BREAKING CHANGE: `model.Position#offset` and `view.Position#offset` now have only a getter. To modify offset use `Postion#getShiftedBy()` or `Position#getShiftedTo()`.

BREAKING CHANGE: `model.Position#path` and `view.Position#path` now have only a getter and are immutable.

BREAKING CHANGE: `model.Range#start` and `model.Range#end` properties now have only a getter.

BREAKING CHANGE: `view.Range#start` and `view.Range#end` properties now have only a getter.

---

### Additional information

This PR required minor changes in other repos (centralized in: https://github.com/ckeditor/ckeditor5/tree/t/ckeditor5-engine/897)
- https://github.com/ckeditor/ckeditor5-list/tree/t/ckeditor5-engine/897
- https://github.com/ckeditor/ckeditor5-ui/tree/t/ckeditor5-engine/897
- https://github.com/ckeditor/ckeditor5-undo/tree/t/ckeditor5-engine/897

Merge commit message for related branches:

```
Internal: Aligned Positions and Ranges usage to the latest API of the framework (see ckeditor/ckeditor5-engine#897).
```

Other:
- Main idea behid this PR is to reduce the need for creating new `Position`/`Range` objects when using it in other methods. Now those object can be safely passed around without side effects.
- The `remove()` method in `view/writer` still does something with passed

scofalik commented on src/dev-utils/view.js:
```diff
@@ -550,7 +550,7 @@ class RangeParser {
 			if ( item.bracket == ELEMENT_RANGE_START_TOKEN || item.bracket == TEXT_RANGE_START_TOKEN ) {
```
Maybe this part of the code can be refactored to store range start and range end, instead of creating a range only to modify it later.

scofalik commented:
I just started to review this, but I wonder if it will be a positive change at all. AFAICS now positions will be created in some scenarios when they weren't, pretty much a lot in OT.

scofalik commented on src/model/operation/transform.js:
```diff
@@ -179,25 +179,29 @@ const ot = {
 				// This nicely covers the joining simplification we did in the previous step.
-				difference.start = difference.start._getTransformedByDeletion( b.sourcePosition, b.howMany );
-				difference.end = difference.end._getTransformedByDeletion( b.sourcePosition, b.howMany );
+				const range = new Range(
```
Please change this const name to `differenceTransformed`.

scofalik commented on src/model/operation/transform.js:
```diff
@@ -179,25 +179,29 @@ const ot = {
 				// Instead, we calculate the new position of that part of original range.
-				common.start = common.start._getCombined( b.sourcePosition, b.getMovedRangeStart() );
-				common.end = common.end._getCombined( b.sourcePosition, b.getMovedRangeStart() );
+				const range = new Range(
```
`commonTransformed`.

scofalik commented on src/model/position.js:
```diff
@@ -482,26 +513,25 @@ export default class Position {
 		// This position can't be affected if deletion was in a different root.
 		if ( this.root != deletePosition.root ) {
-			return transformed;
+			return this;
```
Are we sure this is not harmful in any way? I know that positions are now immutable but this method, by name, suggests that a new instance is returned. I think we don't have (and maybe won't have) such case in code, but one thing that comes to my mind is storing values in `Map` with `Position`s as keys.

jodator replied:
Yeah.. maybe I've gone too deep into removing new instances. Here might be better to always create a copy.

scofalik commented on src/model/position.js:
```diff
@@ -533,19 +567,17 @@ export default class Position {
 		// This position can't be affected if insertion was in a different root.
 		if ( this.root != insertPosition.root ) {
-			return transformed;
+			return this;
```
Same as above.

scofalik commented on src/model/position.js:
```diff
@@ -459,20 +476,34 @@ export default class Position {
+	}
+
+	/**
 	 * Returns a copy of this position that is updated by removing `howMany` nodes starting from `deletePosition`.
```
Docs might need an update if we won't be returning a copy in every scenario (see comment in code below).

scofalik commented on src/model/position.js:
```diff
@@ -782,6 +818,41 @@ export default class Position {
+// @private
+// @param {Array.<Number>} path Position path. See {@link module:engine/model/position~Position#path}.
+function getOffset( path ) {
+	return last( path ) || 0;
```
Why `|| 0`? Wouldn't `undefined` be better?

Anyway, it is an incorrect situation if `path` has no items.

jodator replied:
I didn't want to change logic behind this AFAIR. It was:
```js
get offset() {
    return last( path ) || 0;
}
```
I don't know if it would have any side effects or not.

scofalik replied:
Weird. Position always has to have an offset :S. Try if anything breaks without it ;).

jodator replied:
@scofalik: I was able to fail a test after removing `|| 0`:
```js
expect( new Position( root, [ false ] ) ).to.have.property( 'offset' ).that.equals( 0 );
expect( new Position( root, [ undefined ] ) ).to.have.property( 'offset' ).that.equals( 0 );
```

But this looks like a bad API use - so we might as well drop `|| 0` check as `position` accepts only non-empty array as path.

scofalik replied:
Yeah, I'd remove `|| 0` and the test.

scofalik commented on src/model/range.js:
```diff
@@ -804,13 +829,13 @@ export default class Range {
 		// We are going to modify the result and we need to return a new instance of Range.
 		// We have to create a copy of the reference range.
-		const result = new this( ref.start, ref.end );
+		let result = new this( ref.start, ref.end );
```
AFAICS, range `result` is not used directly, just its `start` and `end` properties. Maybe we could just go with storing `Position`s and it would be better?

scofalik commented on src/model/treewalker.js:
```diff
@@ -204,33 +204,37 @@ export default class TreeWalker {
 	 */
 	_next() {
 		const previousPosition = this.position;
```
Maybe we can drop `position` variable in this function? It is confusing and I am not really sure it is needed.

scofalik commented on src/model/treewalker.js:
```diff
@@ -278,27 +285,31 @@ export default class TreeWalker {
 	 */
 	_previous() {
```
Same with `position` here.

scofalik commented on src/view/treewalker.js:
```diff
@@ -187,7 +187,7 @@ export default class TreeWalker {
 	 */
 	_next() {
-		let position = Position.createFromPosition( this.position );
+		let position = this.position;
```
Same with `position`.

scofalik commented on src/view/treewalker.js:
```diff
@@ -292,7 +293,7 @@ export default class TreeWalker {
 	 */
 	_previous() {
-		let position = Position.createFromPosition( this.position );
+		let position = this.position;
```
Same with `position`.

scofalik commented:
All in all, good job! @Reinmar can you check what impact does it have on the original issue?

scofalik commented:
Could you make PRs out of commits in other repos?

scofalik commented:
@Reinmar @jodator Original issue described problem with copying selection. Unfortunately, this PR only partially fixes it. `Selection#_pushRange` is still the same. One of the reasons is that we don't want people to push `LiveRange`s to `Selection`. So there is still `Range#createFromRange()` there. What do we do with it?

pjasiun commented on src/model/position.js:
```diff
@@ -782,6 +818,41 @@ export default class Position {
+ * @param {module:engine/model/position~Position} position Position of which path should be modified.
+ * @param {Array.<Number>} path Position path. See {@link module:engine/model/position~Position#path}.
+ */
+export function setPath( position, path ) {
```
I would make this a protected class method.

scofalik replied:
I have mixed feelings about this.

From one point we don't want to mess around with position path. On the other hand, though, we need this or `LivePosition` won't work (there is an exactly same problem with `LiveRange#start` and `LiveRange#end`).

If these are protected methods then we don't have to really refactor some parts of the code. Just use the new protected methods! But is this really the way to go?

So IDK, I am torn. I liked this as a utility function because it is more discouraging to use it. But IDK. Maybe `LiveRange` should be in the same file as `Range`. Then we would just use the symbol defined in `range.js`.

pjasiun replied:
For me, it the matter of convention. We have classes with protected methods exposed for other related classes in modules. For instance: https://github.com/ckeditor/ckeditor5-engine/blob/79da6fb1c0782cb6442527d3f2a50497b89b0985/src/model/operation/insertoperation.js#L84 We already had a talk about exposing protected methods for child classes at some point of designing editor class. We never did it as a separate export.

scofalik replied:
I know what is the purpose of `@protected` tag and how we use it. The question is whether we want to have such `@protected` method at all. Because if we do, we can revert (or rather re-do) many of the changes proposed in this PR (at least in OT and tree walkers as these are used multiple time consecutively).

pjasiun replied:
Now I see that also there is a symbol used for the property. I would go with the `@protected` property (without a symbol) and public getters. We have such pair already in some places in the code. We can even write that the property is exposed only for `LiveRange` and should not be used anywhere.

pjasiun commented on src/model/position.js:
```diff
@@ -13,6 +13,9 @@ import compareArrays from '@ckeditor/ckeditor5-utils/src/comparearrays';
 import Text from './text';
 
+const _path = Symbol( 'path' );
+const _root = Symbol( 'root' );
```
We don't use symbols for private properties. `@private` is enough.

scofalik replied:
It is not really `@private`, it is read-only. Also, we do use `Symbol`s in a couple of places. I think these are fine here.

ma2ciek replied:
But symbols make them private too.

jodator replied:
> But symbols make them private too.

A bit better hidden but also accessible ([more work required](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertySymbols) though...) also not enumerable.

But I get that in CKEditor 5 there is already established convention and using straightforward `this._path` in child classes would be simpler.

I only don like "duplication" on object as the getter `this.path` would just return `this._path`.

pjasiun replied:
Feel free to create a separate ticket to change it, but since we already do it this way, we should follow the convention:

https://github.com/ckeditor/ckeditor5-engine/blob/79da6fb1c0782cb6442527d3f2a50497b89b0985/src/model/rootelement.js#L53-L55

https://github.com/ckeditor/ckeditor5-engine/blob/79da6fb1c0782cb6442527d3f2a50497b89b0985/src/view/selection.js#L107-L119

https://github.com/ckeditor/ckeditor5-engine/blob/79da6fb1c0782cb6442527d3f2a50497b89b0985/src/view/text.js#L58-L60

scofalik replied:
I never liked `this._data` and was wondering why it is this way. I guess I will like to change it after more important subjects are dealt with.

Reinmar commented:
Please don't merge this PR before we'll release alpha.2. It's a bit risky change.

Reinmar commented:
Since alpha.2 is out, please conclude this PR and let me know when we could have a final check on it.

jodator commented:
> * getRanges() {
> 	for ( const range of this._ranges ) {
> 		yield Range.createFromRange( range );
> 	}
> }
> 
> Both in model.Selection and view.Selection?
> 
> Same with getFirstRange and getLastRange.

So probably directe getters are to be created as it would make no sense for `getBlabla` methods?

scofalik commented:
No matter what `getRanges()` will return, it should not let modify selection ranges without using selection API. So:
```javascript
get ranges() {
    return this._ranges;
}
```
Is an incorrect solution if you ask me.

jodator commented:
@scofalik 

```js
return [ ...this._ranges ];
```
?

scofalik commented:
Although it was a generator function before, and some people on this project love their generators...

jodator commented:
OK so what to do with those iterators?

I'll grab 🍿 and let you @pjasiun ⚔️  @scofalik fight over removing or not iterators :)

ps.: I'm for returning new array:

```js
get ranges() {
    return [ ...this._ranges ];
```
to not mess with internal array.

scofalik commented:
I am for returning array, because in this case you always want to handle all of them. If not, there are already `getFirstRange()` and `getLastRange()` methods.

pjasiun commented:
I would go with:
```js
get ranges() {
    return this. _ranges[ Symbol.iterator ]();
}
```

This way you will not create any new object and it will be still not mutable.

jodator commented:
@scofalik I've done with all the suggested changes.

I've also removed `Range.createFromRange()` from `getFirst/LastRange()` and updated docs for those methods.

scofalik commented on src/model/range.js:
```diff
@@ -291,8 +311,11 @@ export default class Range {
 
-			pos.path = pos.path.slice( 0, -1 );
-			pos.offset++;
+			const path = pos.getParentPath();
```
I think this is same as `Position.createAfter( pos.parent )`?

scofalik commented on src/model/range.js:
```diff
@@ -755,9 +791,8 @@ export default class Range {
-		const end = Position.createFromPosition( start );
 
-		return new Range( start, end );
+		return new Range( start, start );
```
Even `new Range( start )` should be fine

scofalik commented on src/model/range.js:
```diff
@@ -804,13 +839,14 @@ export default class Range {
 
 		// 5. Ranges should be checked and glued starting from the range that is closest to the reference range.
 		// Since ranges are sorted, start with the range with index that is closest to reference range index.
 		for ( let i = refIndex - 1; i >= 0; i++ ) {
```
I wonder how the hell this works if the condition is `i >= 0` and `i` is increment in each loop. Seems wrong.

scofalik commented on src/model/selection.js:
```diff
@@ -255,7 +253,7 @@ export default class Selection {
 	getLastPosition() {
 		const lastRange = this.getLastRange();
```
-> `last`

scofalik commented on src/view/position.js:
```diff
@@ -24,20 +27,28 @@ export default class Position {
-		 * @member {Number} module:engine/view/position~Position#offset
-		 */
-		this.offset = offset;
+		this[ _parent ] = parent;
```
Didn't we agreed that this should be `this._parent` and `this._offset`?

scofalik commented on src/view/range.js:
```diff
@@ -23,19 +26,8 @@ export default class Range {
-		 * @member {module:engine/view/position~Position}
-		 */
-		this.end = end ? Position.createFromPosition( end ) : Position.createFromPosition( start );
+		this[ _start ] = start;
```
`this._parent` and `this._offset`

scofalik commented on src/view/writer.js:
```diff
@@ -353,9 +352,7 @@ export function remove( range ) {
-	const mergePosition = mergeAttributes( breakStart );
-	range.start = mergePosition;
-	range.end = Position.createFromPosition( mergePosition );
+	mergeAttributes( breakStart );
```
Note to self: check how `mergeAttributres` works now.

scofalik commented:
I've pushed some minor changes.

scofalik commented:
Hm, for some reason it broke: https://travis-ci.org/ckeditor/ckeditor5-engine/builds/306753449?utm_source=email&utm_medium=notification although tests pass for engine for me locally.

Reinmar commented:
I had to revert this change because it completely blows up editor tests.

Also, see https://github.com/ckeditor/ckeditor5-engine/issues/897#issuecomment-346862873.

scofalik commented:
What do you mean exactly by "blows up editor tests"? Performance issues or something else?

Reinmar commented:
> What do you mean exactly by "blows up editor tests"? Performance issues or something else?

Many failing tests in lists, images and undo.
